### PR TITLE
feat: Add branch option for postUpgradeCommands

### LINKED ISSUE #8220: Post upgrade tasks are now being executed for every dependency in a upgrade
**What Renovate type, platform and version are you using?**
Self-hosted **24.11.6** 

**Describe the bug**
Post upgrade tasks are now being executed for every dependency, introduced in: https://github.com/renovatebot/renovate/pull/7735#issuecomment-734274687

We execute codemods shipped with dependencies when upgrading, we run these codemods as 1 command.
This command used to run just once before files were committed, but since #7735 we see the command executed for every dependency.

As a result our codemods are now executed multiple times on the repo when creating a grouped upgrade PR. This makes the renovate build take considerably longer, but also a lot of times results in broken code as not all our codemods are idempotent. What is also worrying is that this change of behaviour went out in a patch, where-as it's a considerable departure of the original behaviour.

Our codemod tool supports targeting specific packages and versions, so we switched our config to the new templating but this surfaces the next bug, the post upgrade tasks are executed for every dependency upgrade for every workspace package. I.e. our mono-repo has 5 listed dependencies on `@atlaskit/icon`, as a result I can see 5 executions of `npx @atlaskit/codemod-cli --packages @atlaskit/icon@21.1.4 --extensions ts,tsx --parser tsx next/packages`, which again means the codemods are executed too many times.

**Have you created a minimal reproduction repository?**

Not sure if this is necessary? The bug is traceable back to the behaviour change introduced in #7735. I'll tick I don't have time for that now, but if asked I'll create a repo.

- [ ] This is a really small bug, it does not need a reproduction (think small typo)
- [ ] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [x] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Proposed solution**
I propose that we support both postUpgradeTasks behaviours and give the config a "executionMode" setting, which can be "per-package" or "per-upgrade". The per-package option would be the new behaviour, per-upgrade the old.

For reference our renovate config:
```
{
  "packageRules": [
    {
      "packagePatterns": "^@confluence/.*",
      "enabled": false
    },
    {
      "groupName": "Atlassian-frontend Scheduled Releases Upgrade",
      "packageNames": [
        "@atlaskit/analytics-next",
        "@atlaskit/css-reset",
        "@atlaskit/reduced-ui-pack",
        "@atlaskit/avatar",
        "@atlaskit/avatar-group",
        "@atlaskit/badge",
        "@atlaskit/banner",
        "@atlaskit/blanket",
        "@atlaskit/breadcrumbs",
        "@atlaskit/button",
        "@atlaskit/calendar",
        "@atlaskit/checkbox",
        "@atlaskit/code",
        "@atlaskit/comment",
        "@atlaskit/datetime-picker",
        "@atlaskit/drawer",
        "@atlaskit/dropdown-menu",
        "@atlaskit/droplist",
        "@atlaskit/dynamic-table",
        "@atlaskit/empty-state",
        "@atlaskit/feedback-collector",
        "@atlaskit/field-base",
        "@atlaskit/field-radio-group",
        "@atlaskit/field-range",
        "@atlaskit/field-text",
        "@atlaskit/field-text-area",
        "@atlaskit/flag",
        "@atlaskit/form",
        "@atlaskit/global-navigation",
        "@atlaskit/icon",
        "@atlaskit/icon-file-type",
        "@atlaskit/icon-object",
        "@atlaskit/icon-priority",
        "@atlaskit/inline-dialog",
        "@atlaskit/inline-edit",
        "@atlaskit/inline-message",
        "@atlaskit/input",
        "@atlaskit/item",
        "@atlaskit/layer",
        "@atlaskit/layer-manager",
        "@atlaskit/logo",
        "@atlaskit/lozenge",
        "@atlaskit/modal-dialog",
        "@atlaskit/motion",
        "@atlaskit/multi-select",
        "@atlaskit/navigation-next",
        "@atlaskit/nps",
        "@atlaskit/onboarding",
        "@atlaskit/page",
        "@atlaskit/page-header",
        "@atlaskit/pagination",
        "@atlaskit/popper",
        "@atlaskit/portal",
        "@atlaskit/progress-bar",
        "@atlaskit/progress-indicator",
        "@atlaskit/progress-tracker",
        "@atlaskit/radio",
        "@atlaskit/range",
        "@atlaskit/rating",
        "@atlaskit/section-message",
        "@atlaskit/select",
        "@atlaskit/single-select",
        "@atlaskit/size-detector",
        "@atlaskit/spinner",
        "@atlaskit/table-tree",
        "@atlaskit/tabs",
        "@atlaskit/tag",
        "@atlaskit/tag-group",
        "@atlaskit/textarea",
        "@atlaskit/textfield",
        "@atlaskit/theme",
        "@atlaskit/toggle",
        "@atlaskit/tooltip",
        "@atlaskit/activity-provider",
        "@atlaskit/adf-schema",
        "@atlaskit/adf-utils",
        "@atlaskit/collab-provider",
        "@atlaskit/editor-bitbucket-transformer",
        "@atlaskit/editor-common",
        "@atlaskit/editor-confluence-transformer",
        "@atlaskit/editor-core",
        "@atlaskit/editor-jira-transformer",
        "@atlaskit/editor-json-transformer",
        "@atlaskit/editor-markdown-transformer",
        "@atlaskit/editor-mobile-bridge",
        "@atlaskit/editor-shared-styles",
        "@atlaskit/email-renderer",
        "@atlaskit/embedded-document",
        "@atlaskit/editor-extension-dropbox",
        "@atlaskit/renderer",
        "@atlaskit/updater-cli",
        "@atlaskit/date",
        "@atlaskit/emoji",
        "@atlaskit/mention",
        "@atlaskit/reactions",
        "@atlaskit/status",
        "@atlaskit/task-decision",
        "@atlaskit/locale",
        "@atlaskit/width-detector",
        "@atlaskit/chunkinator",
        "@atlaskit/media-avatar-picker",
        "@atlaskit/media-card",
        "@atlaskit/media-client",
        "@atlaskit/media-common",
        "@atlaskit/media-core",
        "@atlaskit/media-editor",
        "@atlaskit/media-filmstrip",
        "@atlaskit/media-image",
        "@atlaskit/media-picker",
        "@atlaskit/media-table",
        "@atlaskit/media-test-helpers",
        "@atlaskit/media-ui",
        "@atlaskit/media-viewer",
        "@atlaskit/smart-card",
        "@atlaskit/atlassian-navigation",
        "@atlaskit/atlassian-notifications",
        "@atlaskit/focused-task-close-account",
        "@atlaskit/profilecard"
      ],
      "enabled": true,
      "excludePackagePatterns": [],
      "excludePackageNames": [],
      "reviewers": [
      ]
    }
  ],
  "branchPrefix": "renovate-",
  "extends": [
    "config:base",
    "@atlaskit",
    "@atlaskit:disable-other-upgrades-except-af"
  ],
  "separateMajorMinor": false,
  "rangeStrategy": "bump",
  "prConcurrentLimit": 100,
  "postUpdateOptions": [ "yarnDedupeFewer"],
  "postUpgradeTasks": {
    "commands": [
      "npx @atlaskit/codemod-cli --packages {{{depName}}}@{{{fromVersion}}} --extensions ts,tsx --parser tsx next/packages",
      "npx @atlaskit/codemod-cli --packages {{{depName}}}@{{{fromVersion}}} --extensions ts,tsx --parser tsx packages",
      "npx @atlaskit/codemod-cli --packages {{{depName}}}@{{{fromVersion}}} --extensions js --parser babel next/packages",
      "npx @atlaskit/codemod-cli --packages {{{depName}}}@{{{fromVersion}}} --extensions js --parser babel packages"
    ],
    "fileFilters": ["**/*.{ts,tsx,js}"]
  }
}
```

### PR DESCRIPTION
**Pull request by marcodejongh**
## Changes:

Returns the old postUpgradeTask behaviour under a extra flag
Fixes #8220 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by marcodejongh** [comment]
@sarunint Would you mind helping me on this PR?

File: lib/workers/branch/index.ts
Comment (viceice) [inline lib/workers/branch/index.ts:347]: @rarkins I think this should moved /refacted  to a new file before this pr to be more maintainable. This file is already very big.

**Comment by marcodejongh** [comment]
@viceice Yeah for sure it needs to be abstracted, only reason I didn't straight away is because I wanted to make sure this change worked. I've tested it locally and it works, so I'll work on abstracting it now to get this into a mergable state later this week

File: lib/config/definitions.ts
Comment (rarkins) [inline lib/config/definitions.ts:148]:
```suggestion
    allowedValues: ['update', 'branch'],
```

It's more correctly per-update instead of per-dependency. Potentially there are multiple updates per package within the same branch.

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by kenotron** [comment]
@marcodejongh - very much appreciate this work. An example of usage for me is to execute a bit of logic with https://microsoft.github.io/beachball/ to generate a change file for later automated semver bumping of package versions.

**Comment by marcodejongh** [comment]
@rarkins @viceice I've updated the PR and pulled the post upgrade command executor to a separate file, considering how intertwined it's functionality is with the regular branch worker, I kept the tests in the branch worker itself.

**Comment by marcodejongh** [comment]
@sarunint Would you mind running this branch on your codebase which I assume depends on the behavior change you made?

We'll run the newly added branch mode on our test repository

File: docs/usage/self-hosted-configuration.md
Comment (viceice) [inline docs/usage/self-hosted-configuration.md:12]:
```suggestion
If true allow templating for post-upgrade commands on dependency level post upgrade commands.
```

😉

File: lib/workers/branch/index.spec.ts
Comment (viceice) [inline lib/workers/branch/index.spec.ts:1030]: please use `partial<BranchUpgradeConfig>(...)` from `/test/utils`. So we have more type safety. So we don't need a never cast here.

File: lib/workers/branch/index.ts
Comment (viceice) [inline lib/workers/branch/index.ts:364]: Move this to simple if else in `postUpgradeCommandExecutor`. Why do we need two calls here?

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/workers/branch/index.spec.ts
Comment (marcodejongh) [reply lib/workers/branch/index.spec.ts:1030]: Yeah I copy-pasted this from the other cases, have updated the once I touched

**Comment by sarunint** [comment]
> @sarunint Would you mind running this branch on your codebase which I assume depends on the behavior change you made?
> 
> We'll run the newly added branch mode on our test repository

Sure, I'll test it later today.

File: lib/workers/branch/index.ts
Comment (marcodejongh) [reply lib/workers/branch/index.ts:364]: The looping logic inside `postUpgradeCommandExecutor` becomes a bit more complicated if it's moved inside.
But I'll move it next time I have time to work on this PR

Comment (marcodejongh) [reply lib/workers/branch/index.ts]: Okay actually wasn't too bad at all, pushed now should be good to go 👍

File: lib/workers/branch/index.spec.ts
Comment (marcodejongh) [reply lib/workers/branch/index.spec.ts]: Done, also fixed. There was 1 as never I couldn't remove because the test started failing, someone should probably investigate that. Since it's a test completely unrelated to my changes I've not tried myself.

**Comment by marcodejongh** [comment]
Oops didn't intend to force push, but I didn't know how the eslint pre commit works in this repo.
All feedback has been addressed and the PR should be ready for another review

File: lib/workers/branch/index.ts
Comment (viceice) [inline lib/workers/branch/index.ts:357]: Didn't this override any artifacts already generated by other artifacts functions?

File: lib/workers/branch/execute-post-upgrade-commands.ts
Comment (viceice) [inline lib/workers/branch/execute-post-upgrade-commands.ts:11]:
```suggestion
import type { BranchConfig, BranchUpgradeConfig } from '../types';
```

type only import, will be removed at runtime

File: lib/workers/branch/index.ts
Comment (marcodejongh) [reply lib/workers/branch/index.ts:357]: Inside `await executePostUpgradeCommands(config);` it copies the existing `config.updatedArtifacts` and then iterates over it. The only difference between old and new is that it no longer depends on in-memory editing of the updatedArtifacts object, instead it creates a copy at the start of the run and returns that.

This is also why I override the `updatedArtifacts` in the config passed to the 2nd call of `postUpgradeCommandExecutor`. i.e:

```
  const updatedArtifacts = await postUpgradeCommandExecutor(
    updateUpgradeCommands,
    config,
    getAdminConfig().allowPostUpgradeCommandTemplating
  );
  return postUpgradeCommandExecutor(
    branchUpgradeCommands,
    {
      ...config,
      updatedArtifacts,
    },
    false
  );
```

File: lib/workers/branch/execute-post-upgrade-commands.ts
Comment (marcodejongh) [reply lib/workers/branch/execute-post-upgrade-commands.ts:11]: Done

File: lib/workers/branch/index.ts
Comment (marcodejongh) [reply lib/workers/branch/index.ts]: On closer inspection of the code after replying to you I noticed that I'm not returning artifactErrors, so went and fixed that now it's:

```
return postUpgradeCommandsExecutor(
    branchUpgradeCommands,
    {
      ...config,
      updatedArtifacts,
      artifactErrors,
    },
    false
  );
```

**Comment by Himani-relan** [comment]
Hi, when is this feature expected to release ?

**Review by viceice — COMMENTED** [review]
needs deconflicting

File: lib/workers/branch/execute-post-upgrade-commands.ts
Comment (viceice) [inline lib/workers/branch/execute-post-upgrade-commands.ts:5]:
```suggestion
import type { ArtifactError } from '../../manager/types';
```

File: lib/workers/branch/execute-post-upgrade-commands.ts
Comment (viceice) [inline lib/workers/branch/execute-post-upgrade-commands.ts:14]:
```suggestion
export type PostUpgradeCommandsExecutionResult = {
```

File: lib/workers/branch/execute-post-upgrade-commands.ts
Comment (viceice) [inline lib/workers/branch/execute-post-upgrade-commands.ts:26]:
```suggestion
  const { allowedPostUpgradeCommands, allowPostUpgradeCommandTemplating } = getAdminConfig();
```

Looks cleaner 😉

File: lib/workers/pr/index.ts
Comment (viceice) [inline lib/workers/pr/index.ts:114]:
```suggestion
export type EnsurePrResult = {
```

File: lib/workers/pr/index.ts
Comment (viceice) [inline lib/workers/pr/index.ts:119]:
```suggestion
export async function ensurePr(prConfig: BranchConfig): Promise<EnsurePrResult> {
```

File: lib/workers/branch/index.ts
Comment (viceice) [inline lib/workers/branch/index.ts:354]: @rarkins Should we also remove this here?
```suggestion
```

File: lib/workers/branch/execute-post-upgrade-commands.ts
Comment (viceice) [inline lib/workers/branch/execute-post-upgrade-commands.ts:21]:
```suggestion
```

File: lib/workers/branch/execute-post-upgrade-commands.ts
Comment (viceice) [inline lib/workers/branch/execute-post-upgrade-commands.ts:177]:
```suggestion
```

File: lib/workers/branch/execute-post-upgrade-commands.ts
Comment (viceice) [inline lib/workers/branch/execute-post-upgrade-commands.ts:17]:
```suggestion
};

```

File: lib/workers/branch/execute-post-upgrade-commands.ts
Comment (viceice) [inline lib/workers/branch/execute-post-upgrade-commands.ts:60]:
```suggestion
              ? compile(cmd, upgrade)
```

File: lib/workers/branch/execute-post-upgrade-commands.ts
Comment (viceice) [inline lib/workers/branch/execute-post-upgrade-commands.ts:11]:
```suggestion
import { compile } from '../../util/template';
```

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/workers/branch/index.ts
Comment (marcodejongh) [reply lib/workers/branch/index.ts]: Done, I moved all the "Should postupgradecommands run" logic into `lib/workers/branch/execute-post-upgrade-commands.ts`

**Comment by marcodejongh** [comment]
>  run once instead of running for all dependency upgrades in a single PR.

Correct, that's exactly what we use it for too. In our case, we want to run codemods shipped with the dependencies in the upgrade PR. Theoretically, We could have used per-dependency running for that too, but jscodeshift and `npx` have quite a lot of overhead, which make it impractically in-practice (we saw 5hr runs for 1 upgrade PR when the postUpgradeCommands were first changed)

File: lib/workers/branch/execute-post-upgrade-commands.ts
Comment (viceice) [inline lib/workers/branch/execute-post-upgrade-commands.ts:163]:
```suggestion
```

I think if you already set `allowedPostUpgradeCommands` admin config we do not also require `trustLevel`

File: lib/workers/pr/index.ts
Comment (viceice) [inline lib/workers/pr/index.ts:117]:
```suggestion
};

```

**Review by viceice — COMMENTED** [review]
just  small changes

**Comment by viceice** [comment]
image
😏

File: lib/workers/branch/execute-post-upgrade-commands.ts
Comment (marcodejongh) [reply lib/workers/branch/execute-post-upgrade-commands.ts:163]: Hmmm are you sure? I global searched the repo and I didn't see any checks for `trustLevel` and `allowedPostUpgradeCommands`.
For now I've cleaned up the statement, no need to call getAdminConfig twice

File: lib/workers/branch/execute-post-upgrade-commands.ts
Comment (viceice) [inline lib/workers/branch/execute-post-upgrade-commands.ts:163]:
```suggestion
```

There was a feature request to not need `trustLevel=high` for post upgrade tasks. So this check will be removed in v25 anyways.

`trustLevel` is used if you like to pass full env to child process, with is not always needed and i do not want to allow high trust on  post upgrade tasks 😉

File: lib/workers/branch/execute-post-upgrade-commands.ts
Comment (viceice) [inline lib/workers/branch/execute-post-upgrade-commands.ts:154]:
```suggestion
  const { allowedPostUpgradeCommands } = getAdminConfig();
```

File: lib/workers/pr/index.ts
Comment (viceice) [reply lib/workers/pr/index.ts:117]: Please readd the newline for readabillity.

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by viceice** [comment]
image
😉

File: lib/workers/branch/index.spec.ts
Comment (viceice) [inline lib/workers/branch/index.spec.ts:9]:
```suggestion
import type { WriteExistingFilesResult } from '../../manager/npm/post-update';
```

File: lib/workers/branch/index.spec.ts
Comment (viceice) [inline lib/workers/branch/index.spec.ts:18]:
```suggestion
import type {
```

File: lib/workers/branch/index.spec.ts
Comment (viceice) [inline lib/workers/branch/index.spec.ts:17]:
```suggestion
import type { Pr } from '../repository/onboarding/branch/check';
```

File: lib/workers/branch/index.spec.ts
Comment (viceice) [inline lib/workers/branch/index.spec.ts:28]:
```suggestion
import type { PackageFilesResult } from './get-updated';
```

File: lib/workers/branch/index.spec.ts
Comment (viceice) [inline lib/workers/branch/index.spec.ts:1074]: Is the cast required? shouldn't it work without?

**Review by viceice — COMMENTED** [review]
Please check and remove all casts which are not required by typescript (which you've changed).

File: lib/workers/branch/index.spec.ts
Comment (marcodejongh) [reply lib/workers/branch/index.spec.ts:1074]: Tried that, but this happens: https://imgur.com/a/78GSEJS

Comment (marcodejongh) [reply lib/workers/branch/index.spec.ts:1070]: It's been a while, but I think I tried removing the `as` for all the lines I updated first, if that failed I'd add it back

**Review by viceice — APPROVED** [review]

**Comment by viceice** [comment]
@rarkins needs your approval

**Review by rarkins — APPROVED** [review]

**Comment by viceice** [comment]
@marcodejongh We can't update your branch, please use the update button, otherwise i'm not able to merge.

**Comment by marcodejongh** [comment]
@viceice Hmmm I can't merge it myself either? Maybe something is setup wrong on our atlassian-forks org

**Comment by marcodejongh** [comment]
Or actually I think it would be a permission on this repo that is preventing merging. In the Github UI "base branch" is the target branch

**Comment by viceice** [comment]
The branch update feature does not work cross orgs, it only works for personal accounts.

An alternate is to give us write permission to your fork, so we can update your branch if required.

**Pull request merged**
